### ISSUE TITLE
Add configurable lint for the `Any` type on arbitrary arguments and keyword arguments

### ISSUE THREAD

**Issue by ofek** [astral-sh/ruff#13516]
It would be fantastic to have a rule that prevented `Any` from being used. I would never use this by itself, which is why I'm asking for an option on that rule to only allow that type when the parameters are spelled exactly `*args` and/or `**kwargs`.

Although it's possible to annotate using the [inner type](https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values), and more recently the ability to [`Unpack`](https://peps.python.org/pep-0692/) a `TypedDict`, there are many common situations where one would desire `Any`:

- When the arguments are coming from deserialized data across the wire or user input.
- When overriding a method with new arguments. This saves significant boilerplate and potentially bugs that would be caused by redefining the base class’ method signature.
- When adding annotations to large existing code bases.

If you don't think it would be useful to prevent `Any` ubiquitously then you could just add a rule for the use case that I mentioned.

It's a convention to signal that the arguments should be typed when they are not named exactly `*args`/`**kwargs`. Here's an example from the standard library:

https://github.com/python/cpython/blob/ea5c6507044db0a92f12b5022f91885034fb9a09/Lib/pathlib.py#L711-L717

```python
    def joinpath(self, *pathsegments):
        """Combine this path with one or several arguments, and return a
        new path representing either a subpath (if all arguments are relative
        paths) or a totally different path (if one of the arguments is
        anchored).
        """
        return self.with_segments(self, *pathsegments)
```

**Comment by charliermarsh** [comment]
I think this is somewhat covered by https://docs.astral.sh/ruff/rules/any-type/ -- can you take a look at that rule?

**Comment by ofek** [comment]
Ah I didn't see that one, thanks! As I mentioned though if it existed (which it does) I wouldn't use it without the option I'm asking for. Are you saying that it automatically allows `Any` if the arguments are spelled `*args`/`**kwargs`?

**Comment by charliermarsh** [comment]
I think there's an option for that: https://docs.astral.sh/ruff/settings/#lintflake8-annotations

**Comment by ofek** [comment]
Not quite as that disables ubiquitously for star arguments. Using the example from the standard library, I want this to fail:

```python
def joinpath(self, *pathsegments: Any): ...
```

but this to pass for allowing the situations I enumerated in the OP:

```python
def joinpath(self, *args: Any): ...
```

**Comment by zanieb** [comment]
Can you show this convention used in typed code? My understanding is that the standard library doesn't use type annotations so it doesn't seem like a good example.

**Comment by ofek** [comment]
- https://github.com/psf/black/blob/8d9d18c0338d9fbfc27b38f6f69bb630d90512fa/src/black/output.py#L16 (https://click.palletsprojects.com/en/8.1.x/api/#click.style)
- https://github.com/psf/black/blob/8d9d18c0338d9fbfc27b38f6f69bb630d90512fa/src/black/output.py#L113

**Comment by ofek** [comment]
Another example:

- https://github.com/litestar-org/litestar/blob/b18774922fedc86089d143d9a5484f393826557d/litestar/exceptions/base_exceptions.py#L13
- https://github.com/litestar-org/litestar/blob/b18774922fedc86089d143d9a5484f393826557d/litestar/events/listener.py#L24

**Comment by ofek** [comment]
Another example from the Microsoft cloud SDK: https://github.com/Azure/azure-sdk-for-python/blob/f0c6e8b007699a8c83d7b21d85b1b54621dfd001/sdk/core/corehttp/corehttp/credentials.py#L27

They do that everywhere and it's actually a pattern I forgot to list in the OP. To avoid eventual breaking changes you allow functions to have arbitrary keyword arguments, in which case you don't know the future so you have to use `Any`.

**Comment by MichaReiser** [comment]
CC: @AlexWaygood

**Comment by ofek** [comment]
Has there been any progress on this?
